### PR TITLE
Split forge tests, so they can be filtered out

### PR DESCRIPTION
**Pull request by THenry14**
## Introduced changes

- split e2e and integration tests, so they can now be invoked separately:
```
cargo test integration --manifest-path ./starknet-foundry/Cargo.toml -p forge
cargo test e2e --manifest-path ./starknet-foundry/Cargo.toml -p forge
```
of course running all tests is still possible using the same command as prviously

- Run e2e tests on separate runner, to decrease overall forge tests execution time

## Breaking changes

None

## Checklist

- [X] Linked relevant issue
- [X] Updated relevant documentation
- [X] Added relevant tests
- [X] Performed self-review of the code

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -89,7 +110,7 @@ jobs:
   clippy:
     runs-on: ubuntu-latest
     env:
-      # Make sure CI fails on all warnings, including Clippy lints.
+      # Make sure CI fails on all warnings - including Clippy lints.
```
Comment (THenry14) [inline .github/workflows/ci.yml:113]: this was just to trigger workflow with generated cach, don't mind this

Comment (Arcticae) [reply .github/workflows/ci.yml:113]: You can always re-run manually BTW, and it should have the cache

**Review by Arcticae — APPROVED** [review]
Great idea! :)
This is kind of temporary measure and we perhaps consider more automatic solution, or a pattern of splitting those (cairo-lang has high granularity crates and the tests are blazingly fast :P)

**Comment by THenry14** [comment]
> This is kind of temporary measure and we perhaps consider more automatic solution, or a pattern of splitting those (cairo-lang has high granularity crates and the tests are blazingly fast :P)

100%, this is just temp solution :+1:

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -26,7 +26,28 @@ jobs:
       - uses: software-mansion/setup-scarb@v1.0.0
         with:
           scarb-version: ${{ env.SCARB_VERSION }}
-      - run: cargo test --manifest-path ./starknet-foundry/Cargo.toml -p forge
+      - run: cargo test --lib --manifest-path ./starknet-foundry/Cargo.toml -p forge
```
Comment (cptartur) [inline .github/workflows/ci.yml:29]: Shouldn't this command be removed?

**Review by cptartur — COMMENTED** [review]
If it works, I think we can merge it. 

Though I'm unsure about one thing: Each file in `test` directory is treated as separate create, I wonder if grouping them like that changes this behavior.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -8,8 +8,8 @@ on:
       - master
 
 jobs:
-  test-forge:
-    name: Test Forge
+  test-forge-unit-and-integration:
```
Comment (karol-bisztyga) [inline .github/workflows/ci.yml:11]: hmm, why unit and integration tests are together? couldn't we separate them into two jobs?

**Review by karol-bisztyga — APPROVED** [review]

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]
I think we should add information to readme

File: .github/workflows/ci.yml (hunk 3 above)
Comment (piotmag769) [reply .github/workflows/ci.yml:11]: Units tests are really fast so I don't think we benefit a lot from splitting them

File: .github/workflows/ci.yml (hunk 2 above)
Comment (piotmag769) [reply .github/workflows/ci.yml:29]: Isn't it to run unit tests?

**Review by piotmag769 — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (THenry14) [reply .github/workflows/ci.yml:29]: Exactly, --lib just runs unit tests

**Comment by THenry14** [comment]
> I think we should add information to readme

What info exactly do you have in mind? Things changed for CI, nothing changed from user PoV (the tests still can be run the same as previously)

**Comment by THenry14** [comment]
> Though I'm unsure about one thing: Each file in `test` directory is treated as separate create, I wonder if grouping them like that changes this behavior.

not sure about this either - basically I just renamed main to mod, so I think we should be good? but perhaps some better eperts can have a say

**Comment by cptartur** [comment]
> > Though I'm unsure about one thing: Each file in `test` directory is treated as separate create, I wonder if grouping them like that changes this behavior.
> 
> not sure about this either - basically I just renamed main to mod, so I think we should be good? but perhaps some better eperts can have a say

cc @mkaput

**Comment by mkaput** [comment]
I'm not up-to-date with the discussion but from my experiments in Scarb, the best gains I got by a) splitting integration tests into as many test binaries as possible b) extracting shared tests code to a separate private crate c) only running PR tests on Linux and d) using [nextest and its partitioning feature](https://nexte.st/book/partitioning.html) to let it offload running tests on multiple runners.

**Comment by THenry14** [comment]
> I'm not up-to-date with the discussion but from my experiments in Scarb, the best gains I got by a) splitting integration tests into as many test binaries as possible b) extracting shared tests code to a separate private crate c) only running PR tests on Linux and d) using [nextest and its partitioning feature](https://nexte.st/book/partitioning.html) to let it offload running tests on multiple runners.

Cool, I think this all should be doable here. For now though, we were mainly wondering about:



 > Though I'm unsure about one thing: Each file in test directory is treated as separate create, I wonder if grouping them like that changes this behavior.

> not sure about this either - basically I just renamed main to mod, so I think we should be good? but perhaps some better eperts can have a say

any thoughts on that? @mkaput 🙏

**Comment by mkaput** [comment]
Yeah, from what I see here you have coalesced all tests into a single crate there, because there is only one crate root now (a test named `main` defined by `main.rs` file)

**Comment by karol-bisztyga** [comment]
> > I think we should add information to readme
> 
> What info exactly do you have in mind? Things changed for CI, nothing changed from user PoV (the tests still can be run the same as previously)

I was thinking we could add information on how to run integration tests and how to run e2e tests separately, not just altogether using `cargo test`.

**Review by karol-bisztyga — APPROVED** [review]

**Comment by THenry14** [comment]
> > > I think we should add information to readme
> > 
> > 
> > What info exactly do you have in mind? Things changed for CI, nothing changed from user PoV (the tests still can be run the same as previously)
> 
> I was thinking we could add information on how to run integration tests and how to run e2e tests separately, not just altogether using `cargo test`.

Sure, makes sense. Added :+1:.

**Pull request merged**